bin/lib/rust_library_builder.py

```diff
@@ -0,0 +1,453 @@
+        headers = {"Content-Type": "application/json"}
+
+        url = f'{conanserver_url}/hasfailedbefore'
+        request = requests.post(url, data=json.dumps(self.current_buildparameters_obj), headers=headers)
+        if not request.ok:
+            raise RuntimeError(f'Post failure for {url}: {request}')
+        else:
+            response = json.loads(request.content)
+            return response['response']
+
+    def is_already_uploaded(self, buildfolder, source_folder):
+        annotations = self.get_build_annotations(buildfolder)
+
+        if 'commithash' in annotations:
+            commithash = self.get_commit_hash(source_folder)
+
+            return commithash == annotations['commithash']
+        else:
+            return False
+
+    def set_as_uploaded(self, buildfolder, source_folder):
+        conanhash = self.get_conan_hash(buildfolder)
+        if conanhash is None:
+            raise RuntimeError(f'Error determining conan hash in {buildfolder}')
+
+        self.logger.info(f'commithash: {conanhash}')
+
+        annotations = self.get_build_annotations(buildfolder)
+        if 'commithash' not in annotations:
+            self.upload_builds()
+        annotations['commithash'] = self.get_commit_hash(source_folder)
+
+        self.logger.info(annotations)
+
+        headers = {"Content-Type": "application/json", "Authorization": "Bearer " + self.conanserverproxy_token}
+
+        url = f'{conanserver_url}/annotations/{self.libname}/{self.target_name}/{conanhash}'
+        request = requests.post(url, data=json.dumps(annotations), headers=headers)
+        if not request.ok:
+            raise RuntimeError(f'Post failure for {url}: {request}')
+
+    def clone_branch(self, dest):
+        subprocess.check_call(['git', 'clone', '-q', f'{self.buildconfig.domainurl}/{self.buildconfig.repo}.git', dest],
+                                cwd=self.install_context.staging)
+        subprocess.check_call(['git', '-C', dest, 'checkout', '-q', self.target_name],
+                                cwd=self.install_context.staging)
+
+    def downloadlibrary(self, build_folder):
+        source_folder = os.path.join(build_folder, 'source')
+        self.clone_branch(source_folder)
+        return source_folder
+
+    def makebuildfor(self, compiler, options, exe, compiler_type, toolchain, buildos, buildtype, arch, stdver, stdlib,
+                     flagscombination, ld_path):
+        combined_hash = self.makebuildhash(compiler, options, toolchain, buildos, buildtype, arch, stdver, stdlib,
+                                           flagscombination)
+
+        build_folder = os.path.join(self.install_context.staging, combined_hash)
+        if os.path.exists(build_folder):
+            shutil.rmtree(build_folder, ignore_errors=True)
+        os.makedirs(build_folder, exist_ok=True)
+
+        self.logger.debug(f'Buildfolder: {build_folder}')
+
+        source_folder = self.downloadlibrary(build_folder)
```

Comment:
This is a little intense, we always have to git clone the library to determine the commit hash and compare it with that what was uploaded. Can we somehow circumvent that? Perhaps there is a faster method to get the commit hash from a git repo+branch ?

Reply:
I think some fairly simple parsing of `git ls-remote <gitrepo> <ref>` will do what you want:

```
$ git ls-remote git@github.com:mattgodbolt/performance-tuning.git refs/heads/main
fd7aa6e40e06f09595a56907dffa8f9c5e1cfcb0	refs/heads/main
```